### ISSUE TITLE
Figure out a way to eliminate the "slow first load" problem caused by cron run.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1214]
I have quite a few really small D7 websites that could as well have been built as static, non-dynamic, non-CMS. These get very little traffic and they all have the very same problem: because the last visit they get might have been more than 24h ago, when you try to visit them, they take AGES to load. Alright, not ages but 15-20sec is very common. That is because the cron run is triggered by the visit and the site is not "allowed" to load before that finishes. Not many jobs that I can imagine. Only update manager checking for updates and perhaps indexing/caching jobs I guess.

This phenomenon causes very bad first impression to the visitors as well as the site owners. I could set cron to run on the system level with a cronjob and on top of that at "quiet" times where there is low traffic but these sites -being very small and also small budget- aren't on VPS. As I said, they are small sites ...perfect audience match for Backdrop! The only reason I have not switched them to Backdrop already is because they use some contrib modules not ported yet.

Anyways, there's a lot of people out there that I believe face the same issue and I would love it if Backdrop could make the cron a "background", non-blocking process that lets the site load in the meantime. Have we implemented that already perhaps? ...because my website seems to run faaaast on the fresh 1.2.0!!! :heart: 

There is [Background Process](https://www.drupal.org/project/background_process) (used by modules like [Ultimate Cron](https://www.drupal.org/project/ultimate_cron) or [Cache Graceful](https://www.drupal.org/project/cache_graceful)) that people seem to suggest as a possible solution to that issue. Could we implement something like that in its simplest form?

**Comment by quicksketch** [comment]
Thanks for opening this issue @klonos. This has often driven me crazy as well. I think it may be the source of the problem described at https://github.com/backdrop/backdrop-issues/issues/1137.

I know it's possible to continue processing PHP code after delivering a response to the user, e.g. http://stackoverflow.com/questions/15273570/continue-processing-php-after-sending-http-response

But what we do right now is wait to deliver the page until the cron jobs have finished running. Obviously that's a huge waste of time that we could have done in the background. I'll see if it would be possible to apply this concept to Backdrop.

OTOH, you could always solve this problem by setting up a cron job to run regularly, so cron jobs wouldn't need to fallback to the "safety" cron run provided by Backdrop.

**Comment by klonos** [comment]
> OTOH, you could always solve this problem by setting up a cron job to run regularly, so cron jobs wouldn't need to fallback to the "safety" cron run provided by Backdrop.

Is there a way to do this on economy shared hosting? I think that even if you set the frequency to lower values, still the trigger is someone that visits the site. This "unlucky" person will have to cope with the slowness.

> I'll see if it would be possible to apply this concept to Backdrop.

Yes, I think that this would be the best solution ...if it can be implemented.

**Comment by quicksketch** [comment]
> Is there a way to do this on economy shared hosting? I think that even if you set the frequency to lower values, still the trigger is someone that visits the site.

I think so. I was able to set up cron jobs as an option in CPanel back when I used both Bluehost and A2 Hosting. The code seems to indicate that if cron is set up to be run more frequently than the fallback interval, then it will not over use the fallback mechanism.

**Comment by quicksketch** [comment]
In any case, I still think this is a great idea. As new users are the ones most likely to be affected by this problem, it's most important that we solve it.

I filed a PR at https://github.com/backdrop/backdrop/pull/1072 that I think shows tremendous promise. We'll see if it passes tests. It should be possible to create a test that actually ensures that this "background processing" happens after page delivery.

To test this manually in the mean time, you can just open up index.php and add this line at the bottom:

```
sleep(10);
```

That will make pages take 10 seconds at least to load. After applying the patch, the page will not wait these 10 seconds. You can actually do anything you want after the call to `menu_execute_active_handler()`, but nothing visible will happen on the front-end, because the page has already been delivered.

To make things highly-efficient, I restructured the code for saving a page cache entry. Now the page cache entry is generated and returned immediately, but saved after the page has been delivered. Perhaps a small optimization, but it's one less write query (maybe the only one) on a first-time request.

**Comment by quicksketch** [comment]
Updated with tests for checking that HTML is returned prior to background tasks.

This marks a change for shutdown functions that they are unable to output content to the HTML body. I think that's a fine change, as at best it was used for debugging purposes (e.g. devel query log). I've found that modules are capable of using hook_exit() just as well to do such tasks, if they need to output HTML. Otherwise, they will be relegated to only doing processing and logging, which is reasonable and already the case for any production-like purposes.

**Comment by klonos** [comment]
This sounds great! ...a bit complicated to me, but great. And what a prompt response too!!

Did you base this on [Background Process](https://www.drupal.org/project/background_process) or any other existing Drupal module or patch or was it a solution from scratch? I'm asking because if this does not break any APIs, our brethren might wanna implement the same thing (if applicable I mean).

**Comment by quicksketch** [comment]
> Did you base this on Background Process or any other existing Drupal module or patch or was it a solution from scratch

The basic concept of doing PHP processing while returning a result to the user is based on that [StackOverflow post](http://stackoverflow.com/questions/15273570/continue-processing-php-after-sending-http-response). This same technique is used by Background Process in their method called [BackgroundProcess::flush()](http://cgit.drupalcode.org/background_process/tree/background_process.inc#n977).

Background Process seems to be a much more elaborate solution however, made for executing and monitoring lots of background processes. I *think* the approach used there is a combination of what we've done (returning immediately to the user) and also spawning secondary background processes by doing HTTP requests.

So this approach is only one half of the basic implementation done by Background Process. We might consider doing internal HTTP requests to spawn new processes as well, but considering the use-case for core is really pretty limited compared to what Background Process was made for, I think we'll want to leave this more advanced technique to contrib.

> I'm asking because if this does not break any APIs, our brethren might wanna implement the same thing (if applicable I mean).

Yeah this is essentially a new application of an existing technique. It doesn't break any APIs in the normal sense, though it does cause modules that depended on shutdown functions to use a different mechanism (e.g. hook_exit) if they want to output to the page. Because (by design) we don't wait for additional output after the HTML page has been delivered. That's what allows the background processing to continue separate from the response.

This could easily be adapted to Drupal 7. Drupal 8 could use the same idea, but it would be a different implementation as everything is based on the Symfony Request object there.

What's really exciting about this to me is it opens up the possibility of doing "Background fetch" on cached pages, which is a feature that Varnish 4 uses to eliminate the "first time" slow request effect. Take this situation:

- You have page cache turned on with a 5 minute cache.
- A visitor hits node/1 and gets the first time slow page to prime the cache.
- 5 minutes go by. Another user hits node/1. The cache entry is retrieved but then discarded, because it is too old. The page is generated again and returned from scratch.

With background processing, we could instead serve the *stale* page to the second user. You already have a cache entry for that page, it's just possibly a little out of date. So you return the response from the cached version immediately, but then you build the entire page afterwards, in the background. The cache entry gets updated, and then the *next* user gets the fresh copy.

This would practically eliminate *all* slow page requests. Sure, if you cleared your entire page cache, it would be slow for true first-time requests, but after any page had a cache set, it would always be fast on subsequent hits.

**Comment by docwilmot** [comment]
Advanced gobbledygook to me, but sounds great. Go forth and build that please sir.

**Comment by klonos** [comment]
> Advanced gobbledygook to me, but sounds great. Go forth and build that please sir.

Same here.

**Comment by Graham-72** [comment]
This will be most welcome. So many good things with Backdrop!!!

**Comment by klonos** [comment]
Exactly! :heart: Backdrop.

**Comment by serundeputy** [comment]
there are only two hard problems in computer science:

  * naming things
  * invalidating cache
  * off by one errors 

~Geoff :wink:

**Comment by quicksketch** [comment]
Ha, thanks everyone for your encouragement. :+1: 

I updated the PR at https://github.com/backdrop/backdrop/issues/1072.

It looks like the dream of doing "background fetch" is entirely possible! I'm really stoked about this. Not only could we do cron jobs without holding up the page. We can make it so that all hits to a page are almost *always* cached. Pulling a page out of the Varnish book, this works exactly like its background cache:

- Page cache is set at 5 minutes.
- User A hits node/1. Setting the original page cache.
- All users for the next 5 minutes get the cached copy as created by User A.
- After 5 minutes, User B hits node/1. Previously, the page was regenerated and this user gets a relatively slow load. Now, User B gets the cached copy made by User A. *After* delivering the cached page, Backdrop regenerates the cache and updates it.
- Now for the next 5 minutes, all users get the new cached copy as created by User B.

So this means that nobody ever gets a slow page load, with the exception of the first-ever user. Since we probably don't want to keep a cached copy of every URL forever, I set the time limit to keep expired pages to be 2 days. So as long as someone visits a page once every 2 days, it will always load fast.

I added a new checkbox on the Performance page to toggle this setting:

![screenshot30](https://cloud.githubusercontent.com/assets/100206/9950382/3da009da-5d77-11e5-9a62-c6140437fcfa.png)

API-wise, everything is exactly the same. The *only* caveat is that modules shouldn't use the PHP `header()` function on anonymous pages. That's not a terrible problem however, as all modules are supposed to use `backdrop_add_html_header()` instead anyway. Modules that use `header()` will cause a PHP warning to get logged to watchdog.

**Comment by serundeputy** [comment]
this super goodness; thanks @quicksketch :clap:

**Comment by PaulHoza** [comment]
Ahh!  +100 for the advanced gobbledy-gook!

This is fantastic stuff. Thanks!

**Comment by klonos** [comment]
...and an option on top of that as the cherry to the cake. On by default you think + update existing setups?

**Comment by quicksketch** [comment]
> On by default you think + update existing setups?

Yeah, the current update hook turns it on for existing site. Though they have to have the page cache turned on already. The page cache is turned off by default (something I think we may want to reconsider as well).

**Comment by klonos** [comment]
> The page cache is turned off by default (something I think we may want to reconsider as well).

Yeah, after this improvement, I believe it makes sense to switch that to on by default too.

**Comment by quicksketch** [comment]
I think this code is good for 1.3.x inclusion, but it needs a bit more testing. I added tests for checking that background tasks do not hold up the page load, but this new background fetch functionality needs testing, both with the option turned on and off.

**Comment by serundeputy** [comment]
> new background fetch functionality needs testing, both with the option turned on and off

like [siege](https://www.joedog.org/siege-home/) testing?

**Comment by quicksketch** [comment]
> like siege testing?

No, not performance testing (though I would love that eventually). Rather just functionality testing to ensure the background caches are being generated without any errors.

When testing the background cron functionality, I found that we had a handy system_test.module that made a callback at `system/sleep/%` that you could use to just make a long-running process that ran `sleep(x)`, where x was the parameter in the URL. So testing our background functionality we just check `microtime()` twice before and after making a request, and ensure that it takes less time than the long-running process.

This background page cache is even trickier to test. We'll need to somehow hit a URL, wait a bit for the page to expire, hit it again, and check that we still get a cache. Then hit it again and get a new cached page.

**Comment by guidot** [comment]
Hello everyone!
It seems to me there has been an oversight. What happens if User B does not conveniently wait {cache-timeout} minutes?

*    Page cache is set at 5 minutes.
*    User A hits node/1. Setting the original page cache.
*    All users for the next 5 minutes get the cached copy as created by User A __and the first one that hits node/1 *any time* after 5 minutes too__.
*    If User B hits node/1after 10 hours he/she still gets the cached copy made by User A, right?

That would mean the cache timeout is now out of control. :(

**Comment by klonos** [comment]
> If User B hits node/1after 10 hours he/she still gets the cached copy made by User A, right?

That's right. But they will trigger the cache to be updated (in the background) while they are served the cached copy as created by User A. The next user after User B will be served the fresh copy of the cache (triggered by User B, created in the background - not served to User B).

**Comment by klonos** [comment]
> ...while they are served the cached copy as created by User A.

...should have said "*after* they *have been* served the cached copy as created by User A."

**Comment by quicksketch** [comment]
Yep, spot on @guidot and @klonos. There are two things you could do if this was not desired behavior:

- Turn off the background fetch entirely. (Probably recommended for users if this is a problem).
- Or shorten the stale cache lifetime. I've set it to be 2 days by default. As this is a very confusing setting that would be difficult to explain, I did not put an option for it in the UI. This hidden setting can be modified by changing the value in `system.core.json`. Say if you wanted to make sure that background fetch only served pages that were at most 10 minutes out of date, you would set this value to 600 (seconds). The page would have to be visited within that window in order to get the fast (but stale) load.

This is the same way background fetch works in Varnish. Admittedly it works best for two kinds of sites:

- High traffic sites that are constantly getting hit. In which case caches are "fresh enough" that a one-user stale page won't matter.
- Low traffic sites that are relatively static, so it won't matter if users are served a stale page that may be up to 2 days stale.

If your site is exceptionally slow (something we hope won't happen, but it's possible), it would be preferable to have the page load fast rather than fresh. It will help with your Google ranking if the page loads fast when googlebot comes through, otherwise you get penalized for being slow.

But in any case, there are situations where background fetch won't work well, such as sites that are updated frequently but visited rarely. In which case, background fetch can be turned off.

**Comment by quicksketch** [comment]
I'd like to see this in 1.3.0. We just need tests for this new functionality.

**Comment by MrHaroldA** [comment]
I would love this too!

**Comment by quicksketch** [comment]
PR updated at https://github.com/backdrop/backdrop/pull/1072 with more tests. Now background fetch is tested to ensure stale pages can be delivered and that they are not delivered if the background fetch is disabled.

**Comment by quicksketch** [comment]
Phew, okay that was a rough experience, but tests are now passing in the new PR at https://github.com/backdrop/backdrop/pull/1193.

The previous version would pass when running as Apache through mod_php but not when connecting to PHP-FPM. It turns out that `flush()` will exit a PHP script when running as PHP-FPM, which is not documented anywhere that I could find. As Travis-CI uses PHP-FPM, things weren't working so well there. Now with that figured out, everything is passing locally and on Travis. Very nice.

Barring any other surprises, I think this is ready for merge but I'll give it another look tomorrow.

**Comment by MrHaroldA** [comment]
On a side note; IIUC, Drupal is switching to PHP7 for it's test infrastructure because it runs up to 30-40% faster than 5.x when running tests. I'm deeply impressed of PHP7 too; especially when looking at a Drupal Features overview ;)

**Comment by quicksketch** [comment]
I gave this one last read-through and it looks good. I've merged in https://github.com/backdrop/backdrop/pull/1193 into 1.x for 1.3.0.

On upgrades, the background fetch will be enabled if you have the page cache turned on. The post-processing of cron jobs will always be on. This will need a change record as it has an effect on `hook_exit()`: it can no longer output to the page because the page will have already been delivered to the browser.
